core/metacling/src/TClingCallbacks.cxx

```diff
@@ -750,27 +750,47 @@ void TClingCallbacks::TransactionCommitted(const Transaction &T) {
    TCling__UpdateListsOnCommitted(T, m_Interpreter);
 }
 
-// Collect modules and put them into fPendingCxxModules at first run. Interpreter is not yet initialized at first run
+// Collect decls and put them into fPendingClangDecls at first run. Interpreter is not yet initialized at first run
 // but we need to use interpreter services when loading libraries.
 void TClingCallbacks::beforeExecuteTransaction(const Transaction &T) {
 
-  const std::vector<clang::Module*> &modules = T.getClangModules();
+  const std::vector<const clang::Decl*> &decls = T.getClangDecls();
 
   if (fFirstRun) {
-    for (auto M : modules)
-      fPendingCxxModules.push_back(M);
+    for (auto D : decls) {
+      // Check if Decl is used. If not, don't load the library.
+      if (!D->isUsed())
+        return;
+      fPendingClangDecls.push_back(D);
+    }
     return;
   }
 
-  if (!fPendingCxxModules.empty()) {
-    for (auto M : fPendingCxxModules)
+  if (!fPendingClangDecls.empty()) {
+    for (auto D : fPendingClangDecls) {
+      // Check if Decl is used. If not, don't load the library.
+      if (!D->isUsed())
+        return;
+
+      clang::Module *M = D->getOwningModule();
+      M = M->getTopLevelModule();
+
+      // Load corresponding libraries for this c++ module
       TCling__AutoLoadLibraryForModules(M->Name.c_str());
-    fPendingCxxModules.clear();
+    }
+    fPendingClangDecls.clear();
     return;
   }
 
-  for (auto M : modules)
+  for (auto D : decls) {
+    if (!D->isUsed())
+      return;
+
+    clang::Module *M = D->getOwningModule();
+    M = M->getTopLevelModule();
+
     TCling__AutoLoadLibraryForModules(M->Name.c_str());
```

Comment:
Can we assemble a set of unique `clang::Module*` to call `TCling__AutoLoadLibraryForModules` on, instead of calling it for each `Decl`?